--- udns-0.4/Makefile.in.orig	Thu Jan 23 10:45:31 2014
+++ udns-0.4/Makefile.in	Fri Feb 27 21:12:14 2015
@@ -40,6 +40,7 @@
 SOBJS  = $(OBJS:.o=.lo)
 SOLIB  = lib$(NAME)_s.so
 SOLIBV = lib$(NAME).so.$(SOVER)
+SOLIBS = lib$(NAME).so
 SOLIBFL= -L. -l$(NAME)_s
 
 UTILS   = $(USRCS:.c=)
@@ -58,7 +59,15 @@
 PICFLAGS = -fPIC
 AWK = awk
 TAR = tar
+LN_S = ln -sf
+INSTALL = install
 
+PREFIX     = /usr/local
+INCLUDEDIR = $(PREFIX)/include
+LIBDIR     = $(PREFIX)/lib
+BINDIR     = $(PREFIX)/bin
+MANDIR     = $(PREFIX)/share/man
+
 all: static
 
 .SUFFIXES: .c .o .lo
@@ -75,7 +84,7 @@
 sharedlib: $(SOLIBV)
 
 $(SOLIBV): $(SOBJS)
-	$(LDSHARED) -Wl,--soname,$(SOLIBV) -o $@ $(SOBJS) $(LDFLAGS) $(LIBS)
+	$(LDSHARED) -Wl,-h,$(SOLIBV) -o $@ $(SOBJS) $(LDFLAGS) $(LIBS)
 $(SOLIB): $(SOLIBV)
 	rm -f $@
 	ln -s $(SOLIBV) $@
@@ -141,6 +150,23 @@
 .PHONY: all static staticlib shared sharedlib dist clean distclean subdist \
   depend dep deps
 
+install:
+	$(INSTALL) -d $(DESTDIR)$(INCLUDEDIR)
+	$(INSTALL) -m 0444 udns.h $(DESTDIR)$(INCLUDEDIR)/udns.h
+	$(INSTALL) -d $(DESTDIR)$(BINDIR)
+	for i in $(UTILS) ; do \
+		$(INSTALL) -m 0755 $$i $(DESTDIR)$(BINDIR)/$$i ; \
+	done
+	$(INSTALL) -d $(DESTDIR)$(MANDIR)/man1
+	$(INSTALL) -d $(DESTDIR)$(MANDIR)/man3
+	for i in dnsget.1 rblcheck.1 ; do \
+		$(INSTALL) -m 0644 $$i $(DESTDIR)$(MANDIR)/man1/$$i ; \
+	done
+	$(INSTALL) -m 0644 udns.3 $(DESTDIR)$(MANDIR)/man3/udns.3
+	$(INSTALL) -d $(DESTDIR)$(LIBDIR)
+	$(INSTALL) -m 0555 $(SOLIBV) $(DESTDIR)$(LIBDIR)/$(SOLIBV)
+	$(LN_S) $(SOLIBV) $(DESTDIR)$(LIBDIR)/$(SOLIBS)
+
 depend dep deps: $(SRCS) $(USRC)
 	@echo Generating deps for:
 	@echo \ $(SRCS)
@@ -153,7 +179,7 @@
 	done >> Makefile.tmp; \
 	for f in $(USRCS:.c=.o); do \
 	 echo "$${f%.?}: $$f \$$(LIB)"; \
-	 echo "	\$$(LD) \$$(LDLAGS) -o \$$@ $$f \$$(LIBFL) \$$(LIBS)"; \
+	 echo "	\$$(LD) \$$(LDFLAGS) -o \$$@ $$f \$$(LIBFL) \$$(LIBS)"; \
 	 echo "$${f%.?}_s: $$f \$$(SOLIB)"; \
 	 echo "	\$$(LD) \$$(LDFLAGS) -o \$$@ $$f \$$(SOLIBFL)"; \
 	done >> Makefile.tmp ; \
@@ -184,14 +210,14 @@
 rblcheck.o rblcheck.lo: rblcheck.c config.h udns.h getopt.c
 ex-rdns.o ex-rdns.lo: ex-rdns.c udns.h
 dnsget: dnsget.o $(LIB)
-	$(LD) $(LDLAGS) -o $@ dnsget.o $(LIBFL) $(LIBS)
+	$(LD) $(LDFLAGS) -o $@ dnsget.o $(LIBFL) $(LIBS)
 dnsget_s: dnsget.o $(SOLIB)
 	$(LD) $(LDFLAGS) -o $@ dnsget.o $(SOLIBFL)
 rblcheck: rblcheck.o $(LIB)
-	$(LD) $(LDLAGS) -o $@ rblcheck.o $(LIBFL) $(LIBS)
+	$(LD) $(LDFLAGS) -o $@ rblcheck.o $(LIBFL) $(LIBS)
 rblcheck_s: rblcheck.o $(SOLIB)
 	$(LD) $(LDFLAGS) -o $@ rblcheck.o $(SOLIBFL)
 ex-rdns: ex-rdns.o $(LIB)
-	$(LD) $(LDLAGS) -o $@ ex-rdns.o $(LIBFL) $(LIBS)
+	$(LD) $(LDFLAGS) -o $@ ex-rdns.o $(LIBFL) $(LIBS)
 ex-rdns_s: ex-rdns.o $(SOLIB)
 	$(LD) $(LDFLAGS) -o $@ ex-rdns.o $(SOLIBFL)
